File: sdk/core/unity/Runtime/Core/UsageMeter.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace Anthropic.SDK.Metering
{
    /// <summary>
    /// Types of billable usage events
    /// </summary>
    public enum UsageEventType
    {
        AdRequest,
        AdImpression,
        AdClick,
        AdVideoStart,
        AdVideoComplete,
        AdRevenue,
        ApiCall,
        CacheHit,
        CacheMiss,
        Error
    }

    /// <summary>
    /// A single usage event for metering
    /// </summary>
    public struct UsageEvent
    {
        public UsageEventType Type { get; set; }
        public DateTime Timestamp { get; set; }
        public string PlacementId { get; set; }
        public string AdapterId { get; set; }
        public string AdFormat { get; set; }
        public double? RevenueAmount { get; set; }
        public Dictionary<string, object> Metadata { get; set; }

        public static UsageEvent Create(UsageEventType type, string placementId = null, 
            string adapterId = null, string adFormat = null, double? revenueAmount = null)
        {
            return new UsageEvent
            {
                Type = type,
                Timestamp = DateTime.UtcNow,
                PlacementId = placementId,
                AdapterId = adapterId,
                AdFormat = adFormat,
                RevenueAmount = revenueAmount
            };
        }
    }

    /// <summary>
    /// Aggregated usage metrics for a period
    /// </summary>
    public struct UsageMetrics
    {
        public long AdRequests { get; set; }
        public long AdImpressions { get; set; }
        public long AdClicks { get; set; }
        public long VideoStarts { get; set; }
        public long VideoCompletes { get; set; }
        public double TotalRevenue { get; set; }
        public long ApiCalls { get; set; }
        public long CacheHits { get; set; }
        public long CacheMisses { get; set; }
        public long Errors { get; set; }
        public DateTime PeriodStart { get; set; }
        public DateTime PeriodEnd { get; set; }
    }

    /// <summary>
    /// Metrics broken down by dimension
    /// </summary>
    public struct UsageBreakdown
    {
        public Dictionary<string, UsageMetrics> ByPlacement { get; set; }
        public Dictionary<string, UsageMetrics> ByAdapter { get; set; }
        public Dictionary<string, UsageMetrics> ByAdFormat { get; set; }
    }

    /// <summary>
    /// Configuration for the usage meter
    /// </summary>
    public class MeteringConfig
    {
        public TimeSpan FlushInterval { get; set; } = TimeSpan.FromMinutes(1);
        public int MaxEventsBeforeFlush { get; set; } = 1000;
        public bool EnableLocalStorage { get; set; } = true;
        public bool EnableRemoteReporting { get; set; } = true;
        public double SamplingRate { get; set; } = 1.0;
    }

    /// <summary>
    /// Interface for reporting metrics to remote service
    /// </summary>
    public interface IMeteringReporter
    {
        Task<bool> ReportAsync(UsageMetrics metrics, UsageBreakdown breakdown, 
            CancellationToken cancellationToken = default);
    }

    /// <summary>
    /// Dimension-specific metrics
    /// </summary>
    internal class DimensionMetrics
    {
        private long _adRequests;
        private long _adImpressions;
        private long _adClicks;
        private long _videoStarts;
        private long _videoCompletes;
        private double _totalRevenue;
        private long _apiCalls;
        private long _cacheHits;
        private long _cacheMisses;
        private long _errors;
        private readonly object _revenueLock = new object();

        public void IncrementAdRequests() => Interlocked.Increment(ref _adRequests);
        public void IncrementAdImpressions() => Interlocked.Increment(ref _adImpressions);
        public void IncrementAdClicks() => Interlocked.Increment(ref _adClicks);
        public void IncrementVideoStarts() => Interlocked.Increment(ref _videoStarts);
        public void IncrementVideoCompletes() => Interlocked.Increment(ref _videoCompletes);
        public void IncrementApiCalls() => Interlocked.Increment(ref _apiCalls);
        public void IncrementCacheHits() => Interlocked.Increment(ref _cacheHits);
        public void IncrementCacheMisses() => Interlocked.Increment(ref _cacheMisses);
        public void IncrementErrors() => Interlocked.Increment(ref _errors);

        public void AddRevenue(double amount)
        {
            lock (_revenueLock)
            {
                _totalRevenue += amount;
            }
        }

        public UsageMetrics ToMetrics(DateTime periodStart, DateTime periodEnd)
        {
            return new UsageMetrics
            {
                AdRequests = Interlocked.Read(ref _adRequests),
                AdImpressions = Interlocked.Read(ref _adImpressions),
                AdClicks = Interlocked.Read(ref _adClicks),
                VideoStarts = Interlocked.Read(ref _videoStarts),
                VideoCompletes = Interlocked.Read(ref _videoCompletes),
                TotalRevenue = _totalRevenue,
                ApiCalls = Interlocked.Read(ref _apiCalls),
                CacheHits = Interlocked.Read(ref _cacheHits),
                CacheMisses = Interlocked.Read(ref _cacheMisses),
                Errors = Interlocked.Read(ref _errors),
                PeriodStart = periodStart,
                PeriodEnd = periodEnd
            };
        }

        public void Reset()
        {
            Interlocked.Exchange(ref _adRequests, 0);
            Interlocked.Exchange(ref _adImpressions, 0);
            Interlocked.Exchange(ref _adClicks, 0);
            Interlocked.Exchange(ref _videoStarts, 0);
            Interlocked.Exchange(ref _videoCompletes, 0);
            lock (_revenueLock) { _totalRevenue = 0; }
            Interlocked.Exchange(ref _apiCalls, 0);
            Interlocked.Exchange(ref _cacheHits, 0);
            Interlocked.Exchange(ref _cacheMisses, 0);
            Interlocked.Exchange(ref _errors, 0);
        }
    }

    /// <summary>
    /// Tracks and reports billable usage events
    /// </summary>
    public class UsageMeter : IDisposable
    {
        private readonly MeteringConfig _config;
        private readonly IMeteringReporter _reporter;
        
        // Global counters
        private readonly DimensionMetrics _globalMetrics = new DimensionMetrics();
        
        // Dimensional breakdowns
        private readonly ConcurrentDictionary<string, DimensionMetrics> _placementMetrics = new();
        private readonly ConcurrentDictionary<string, DimensionMetrics> _adapterMetrics = new();
        private readonly ConcurrentDictionary<string, DimensionMetrics> _formatMetrics = new();
        
        // Pending events
        private readonly List<UsageEvent> _pendingEvents = new();
        private readonly object _eventsLock = new object();
        
        // Period tracking
        private DateTime _periodStart = DateTime.UtcNow;
        
        // Flush timer
        private readonly CancellationTokenSource _cts = new();
        private Task _flushTask;
        private bool _isRunning;
        private readonly Random _random = new();

        public UsageMeter(MeteringConfig config = null, IMeteringReporter reporter = null)
        {
            _config = config ?? new MeteringConfig();
            _reporter = reporter;
        }

        /// <summary>
        /// Start the metering service
        /// </summary>
        public void Start()
        {
            if (_isRunning) return;
            _isRunning = true;
            _periodStart = DateTime.UtcNow;

            _flushTask = Task.Run(async () =>
            {
                while (!_cts.Token.IsCancellationRequested && _isRunning)
                {
                    await Task.Delay(_config.FlushInterval, _cts.Token).ConfigureAwait(false);
                    await FlushAsync(_cts.Token).ConfigureAwait(false);
                }
            });
        }

        /// <summary>
        /// Stop the metering service
        /// </summary>
        public void Stop()
        {
            _isRunning = false;
            _cts.Cancel();
        }

        /// <summary>
        /// Record a usage event
        /// </summary>
        public void Record(UsageEvent evt)
        {
            // Apply sampling
            if (_config.SamplingRate < 1.0 && _random.NextDouble() > _config.SamplingRate)
            {
                return;
            }

            // Update global counters
            UpdateMetrics(_globalMetrics, evt);

            // Update dimensional breakdowns
            if (!string.IsNullOrEmpty(evt.PlacementId))
            {
                var metrics = _placementMetrics.GetOrAdd(evt.PlacementId, _ => new DimensionMetrics());
                UpdateMetrics(metrics, evt);
            }

            if (!string.IsNullOrEmpty(evt.AdapterId))
            {
                var metrics = _adapterMetrics.GetOrAdd(evt.AdapterId, _ => new DimensionMetrics());
                UpdateMetrics(metrics, evt);
            }

            if (!string.IsNullOrEmpty(evt.AdFormat))
            {
                var metrics = _formatMetrics.GetOrAdd(evt.AdFormat, _ => new DimensionMetrics());
                UpdateMetrics(metrics, evt);
            }

            // Queue event for storage
            if (_config.EnableLocalStorage)
            {
                lock (_eventsLock)
                {
                    _pendingEvents.Add(evt);
                    if (_pendingEvents.Count >= _config.MaxEventsBeforeFlush)
                    {
                        _ = FlushAsync(CancellationToken.None);
                    }
                }
            }
        }

        private void UpdateMetrics(DimensionMetrics metrics, UsageEvent evt)
        {
            switch (evt.Type)
            {
                case UsageEventType.AdRequest: metrics.IncrementAdRequests(); break;
                case UsageEventType.AdImpression: metrics.IncrementAdImpressions(); break;
                case UsageEventType.AdClick: metrics.IncrementAdClicks(); break;
                case UsageEventType.AdVideoStart: metrics.IncrementVideoStarts(); break;
                case UsageEventType.AdVideoComplete: metrics.IncrementVideoCompletes(); break;
                case UsageEventType.AdRevenue:
                    if (evt.RevenueAmount.HasValue) metrics.AddRevenue(evt.RevenueAmount.Value);
                    break;
                case UsageEventType.ApiCall: metrics.IncrementApiCalls(); break;
                case UsageEventType.CacheHit: metrics.IncrementCacheHits(); break;
                case UsageEventType.CacheMiss: metrics.IncrementCacheMisses(); break;
                case UsageEventType.Error: metrics.IncrementErrors(); break;
            }
        }

        // Convenience recording methods
        public void RecordRequest(string placementId, string adapterId = null, string adFormat = null) =>
            Record(UsageEvent.Create(UsageEventType.AdRequest, placementId, adapterId, adFormat));

        public void RecordImpression(string placementId, string adapterId, string adFormat = null) =>
            Record(UsageEvent.Create(UsageEventType.AdImpression, placementId, adapterId, adFormat));

        public void RecordClick(string placementId, string adapterId, string adFormat = null) =>
            Record(UsageEvent.Create(UsageEventType.AdClick, placementId, adapterId, adFormat));

        public void RecordRevenue(string placementId, string adapterId, double amount) =>
            Record(UsageEvent.Create(UsageEventType.AdRevenue, placementId, adapterId, null, amount));

        public void RecordVideoStart(string placementId, string adapterId) =>
            Record(UsageEvent.Create(UsageEventType.AdVideoStart, placementId, adapterId, "video"));

        public void RecordVideoComplete(string placementId, string adapterId) =>
            Record(UsageEvent.Create(UsageEventType.AdVideoComplete, placementId, adapterId, "video"));

        public void RecordCacheHit(string placementId = null) =>
            Record(UsageEvent.Create(UsageEventType.CacheHit, placementId));

        public void RecordCacheMiss(string placementId = null) =>
            Record(UsageEvent.Create(UsageEventType.CacheMiss, placementId));

        public void RecordError(string placementId = null, string adapterId = null) =>
            Record(UsageEvent.Create(UsageEventType.Error, placementId, adapterId));

        /// <summary>
        /// Get current metrics snapshot
        /// </summary>
        public UsageMetrics GetMetrics()
        {
            return _globalMetrics.ToMetrics(_periodStart, DateTime.UtcNow);
        }

        /// <summary>
        /// Get breakdown by dimensions
        /// </summary>
        public UsageBreakdown GetBreakdown()
        {
            var now = DateTime.UtcNow;
            return new UsageBreakdown
            {
                ByPlacement = new Dictionary<string, UsageMetrics>(
                    _placementMetrics.Select(kv => new KeyValuePair<string, UsageMetrics>(
                        kv.Key, kv.Value.ToMetrics(_periodStart, now)))),
                ByAdapter = new Dictionary<string, UsageMetrics>(
                    _adapterMetrics.Select(kv => new KeyValuePair<string, UsageMetrics>(
                        kv.Key, kv.Value.ToMetrics(_periodStart, now)))),
                ByAdFormat = new Dictionary<string, UsageMetrics>(
                    _formatMetrics.Select(kv => new KeyValuePair<string, UsageMetrics>(
                        kv.Key, kv.Value.ToMetrics(_periodStart, now))))
            };
        }

        /// <summary>
        /// Get click-through rate
        /// </summary>
        public double GetCTR()
        {
            var metrics = GetMetrics();
            if (metrics.AdImpressions == 0) return 0;
            return (double)metrics.AdClicks / metrics.AdImpressions;
        }

        /// <summary>
        /// Get fill rate
        /// </summary>
        public double GetFillRate()
        {
            var metrics = GetMetrics();
            if (metrics.AdRequests == 0) return 0;
            return (double)metrics.AdImpressions / metrics.AdRequests;
        }

        /// <summary>
        /// Get video completion rate
        /// </summary>
        public double GetVideoCompletionRate()
        {
            var metrics = GetMetrics();
            if (metrics.VideoStarts == 0) return 0;
            return (double)metrics.VideoCompletes / metrics.VideoStarts;
        }

        /// <summary>
        /// Get cache hit rate
        /// </summary>
        public double GetCacheHitRate()
        {
            var metrics = GetMetrics();
            var total = metrics.CacheHits + metrics.CacheMisses;
            if (total == 0) return 0;
            return (double)metrics.CacheHits / total;
        }

        /// <summary>
        /// Get effective CPM
        /// </summary>
        public double GetEffectiveCPM()
        {
            var metrics = GetMetrics();
            if (metrics.AdImpressions == 0) return 0;
            return (metrics.TotalRevenue / metrics.AdImpressions) * 1000.0;
        }

        /// <summary>
        /// Flush metrics to reporter
        /// </summary>
        public async Task<bool> FlushAsync(CancellationToken cancellationToken = default)
        {
            if (!_config.EnableRemoteReporting || _reporter == null)
            {
                return true;
            }

            var metrics = GetMetrics();
            var breakdown = GetBreakdown();

            try
            {
                var success = await _reporter.ReportAsync(metrics, breakdown, cancellationToken)
                    .ConfigureAwait(false);
                if (success)
                {
                    Reset();
                }
                return success;
            }
            catch
            {
                return false;
            }
        }

        /// <summary>
        /// Reset all counters
        /// </summary>
        public void Reset()
        {
            _globalMetrics.Reset();
            foreach (var metrics in _placementMetrics.Values) metrics.Reset();
            foreach (var metrics in _adapterMetrics.Values) metrics.Reset();
            foreach (var metrics in _formatMetrics.Values) metrics.Reset();

            lock (_eventsLock)
            {
                _pendingEvents.Clear();
            }

            _periodStart = DateTime.UtcNow;
        }

        /// <summary>
        /// Export metrics as JSON
        /// </summary>
        public string ExportAsJSON()
        {
            var metrics = GetMetrics();
            var sb = new StringBuilder();
            sb.AppendLine("{");
            sb.AppendLine("  \"metrics\": {");
            sb.AppendLine($"    \"adRequests\": {metrics.AdRequests},");
            sb.AppendLine($"    \"adImpressions\": {metrics.AdImpressions},");
            sb.AppendLine($"    \"adClicks\": {metrics.AdClicks},");
            sb.AppendLine($"    \"videoStarts\": {metrics.VideoStarts},");
            sb.AppendLine($"    \"videoCompletes\": {metrics.VideoCompletes},");
            sb.AppendLine($"    \"totalRevenue\": {metrics.TotalRevenue},");
            sb.AppendLine($"    \"apiCalls\": {metrics.ApiCalls},");
            sb.AppendLine($"    \"cacheHits\": {metrics.CacheHits},");
            sb.AppendLine($"    \"cacheMisses\": {metrics.CacheMisses},");
            sb.AppendLine($"    \"errors\": {metrics.Errors}");
            sb.AppendLine("  },");
            sb.AppendLine("  \"computed\": {");
            sb.AppendLine($"    \"ctr\": {GetCTR()},");
            sb.AppendLine($"    \"fillRate\": {GetFillRate()},");
            sb.AppendLine($"    \"videoCompletionRate\": {GetVideoCompletionRate()},");
            sb.AppendLine($"    \"cacheHitRate\": {GetCacheHitRate()},");
            sb.AppendLine($"    \"effectiveCPM\": {GetEffectiveCPM()}");
            sb.AppendLine("  }");
            sb.AppendLine("}");
            return sb.ToString();
        }

        public void Dispose()
        {
            Stop();
            _cts.Dispose();
        }
    }

    /// <summary>
    /// Builder for UsageMeter
    /// </summary>
    public class UsageMeterBuilder
    {
        private MeteringConfig _config = new MeteringConfig();
        private IMeteringReporter _reporter;

        public UsageMeterBuilder WithConfig(MeteringConfig config)
        {
            _config = config;
            return this;
        }

        public UsageMeterBuilder WithReporter(IMeteringReporter reporter)
        {
            _reporter = reporter;
            return this;
        }

        public UsageMeterBuilder WithFlushInterval(TimeSpan interval)
        {
            _config.FlushInterval = interval;
            return this;
        }

        public UsageMeterBuilder WithSamplingRate(double rate)
        {
            _config.SamplingRate = Math.Max(0, Math.Min(1, rate));
            return this;
        }

        public UsageMeter Build()
        {
            return new UsageMeter(_config, _reporter);
        }
    }
}

File: sdk/web/src/usageMeter.ts
/**
 * @file Usage Meter
 * @description Track and report billable usage events for ad SDK
 * @module @rivalapex/web-sdk
 */

/**
 * Types of billable usage events
 */
export enum UsageEventType {
  AD_REQUEST = 'ad_request',
  AD_IMPRESSION = 'ad_impression',
  AD_CLICK = 'ad_click',
  AD_VIDEO_START = 'ad_video_start',
  AD_VIDEO_COMPLETE = 'ad_video_complete',
  AD_REVENUE = 'ad_revenue',
  API_CALL = 'api_call',
  CACHE_HIT = 'cache_hit',
  CACHE_MISS = 'cache_miss',
  ERROR = 'error',
}

/**
 * A single usage event
 */
export interface UsageEvent {
  type: UsageEventType;
  timestamp: number;
  placementId?: string;
  adapterId?: string;
  adFormat?: string;
  revenueAmount?: number;
  metadata?: Record<string, unknown>;
}

/**
 * Aggregated usage metrics
 */
export interface UsageMetrics {
  adRequests: number;
  adImpressions: number;
  adClicks: number;
  videoStarts: number;
  videoCompletes: number;
  totalRevenue: number;
  apiCalls: number;
  cacheHits: number;
  cacheMisses: number;
  errors: number;
  periodStart: number;
  periodEnd: number;
}

/**
 * Breakdown by dimension
 */
export interface UsageBreakdown {
  byPlacement: Map<string, UsageMetrics>;
  byAdapter: Map<string, UsageMetrics>;
  byAdFormat: Map<string, UsageMetrics>;
}

/**
 * Metering configuration
 */
export interface MeteringConfig {
  flushIntervalMs?: number;
  maxEventsBeforeFlush?: number;
  enableLocalStorage?: boolean;
  enableRemoteReporting?: boolean;
  samplingRate?: number;
  reportEndpoint?: string;
}

/**
 * Reporter interface for sending metrics
 */
export interface MeteringReporter {
  report(metrics: UsageMetrics, breakdown: UsageBreakdown): Promise<boolean>;
}

/**
 * Dimension-specific metrics
 */
class DimensionMetrics {
  adRequests = 0;
  adImpressions = 0;
  adClicks = 0;
  videoStarts = 0;
  videoCompletes = 0;
  totalRevenue = 0;
  apiCalls = 0;
  cacheHits = 0;
  cacheMisses = 0;
  errors = 0;

  toMetrics(periodStart: number, periodEnd: number): UsageMetrics {
    return {
      adRequests: this.adRequests,
      adImpressions: this.adImpressions,
      adClicks: this.adClicks,
      videoStarts: this.videoStarts,
      videoCompletes: this.videoCompletes,
      totalRevenue: this.totalRevenue,
      apiCalls: this.apiCalls,
      cacheHits: this.cacheHits,
      cacheMisses: this.cacheMisses,
      errors: this.errors,
      periodStart,
      periodEnd,
    };
  }

  reset(): void {
    this.adRequests = 0;
    this.adImpressions = 0;
    this.adClicks = 0;
    this.videoStarts = 0;
    this.videoCompletes = 0;
    this.totalRevenue = 0;
    this.apiCalls = 0;
    this.cacheHits = 0;
    this.cacheMisses = 0;
    this.errors = 0;
  }
}

/**
 * Default configuration
 */
const DEFAULT_CONFIG: Required<MeteringConfig> = {
  flushIntervalMs: 60000,
  maxEventsBeforeFlush: 1000,
  enableLocalStorage: true,
  enableRemoteReporting: true,
  samplingRate: 1.0,
  reportEndpoint: '/api/v1/metrics/usage',
};

/**
 * Usage meter for tracking billable events
 */
export class UsageMeter {
  private config: Required<MeteringConfig>;
  private reporter?: MeteringReporter;

  // Global metrics
  private globalMetrics = new DimensionMetrics();

  // Dimensional breakdowns
  private placementMetrics = new Map<string, DimensionMetrics>();
  private adapterMetrics = new Map<string, DimensionMetrics>();
  private formatMetrics = new Map<string, DimensionMetrics>();

  // Pending events
  private pendingEvents: UsageEvent[] = [];

  // Period tracking
  private periodStart: number = Date.now();

  // Flush timer
  private flushTimer: ReturnType<typeof setInterval> | null = null;
  private isRunning = false;

  constructor(config: MeteringConfig = {}, reporter?: MeteringReporter) {
    this.config = { ...DEFAULT_CONFIG, ...config };
    this.reporter = reporter;
  }

  /**
   * Start the metering service
   */
  start(): void {
    if (this.isRunning) return;
    this.isRunning = true;
    this.periodStart = Date.now();

    this.flushTimer = setInterval(() => {
      this.flush().catch(() => {});
    }, this.config.flushIntervalMs);
  }

  /**
   * Stop the metering service
   */
  stop(): void {
    this.isRunning = false;
    if (this.flushTimer) {
      clearInterval(this.flushTimer);
      this.flushTimer = null;
    }
  }

  /**
   * Record a usage event
   */
  record(event: UsageEvent): void {
    // Apply sampling
    if (this.config.samplingRate < 1.0 && Math.random() > this.config.samplingRate) {
      return;
    }

    // Update global metrics
    this.updateMetrics(this.globalMetrics, event);

    // Update dimensional breakdowns
    if (event.placementId) {
      const metrics = this.getOrCreateMetrics(this.placementMetrics, event.placementId);
      this.updateMetrics(metrics, event);
    }

    if (event.adapterId) {
      const metrics = this.getOrCreateMetrics(this.adapterMetrics, event.adapterId);
      this.updateMetrics(metrics, event);
    }

    if (event.adFormat) {
      const metrics = this.getOrCreateMetrics(this.formatMetrics, event.adFormat);
      this.updateMetrics(metrics, event);
    }

    // Queue event for storage
    if (this.config.enableLocalStorage) {
      this.pendingEvents.push(event);
      if (this.pendingEvents.length >= this.config.maxEventsBeforeFlush) {
        this.flush().catch(() => {});
      }
    }
  }

  private getOrCreateMetrics(map: Map<string, DimensionMetrics>, key: string): DimensionMetrics {
    let metrics = map.get(key);
    if (!metrics) {
      metrics = new DimensionMetrics();
      map.set(key, metrics);
    }
    return metrics;
  }

  private updateMetrics(metrics: DimensionMetrics, event: UsageEvent): void {
    switch (event.type) {
      case UsageEventType.AD_REQUEST:
        metrics.adRequests++;
        break;
      case UsageEventType.AD_IMPRESSION:
        metrics.adImpressions++;
        break;
      case UsageEventType.AD_CLICK:
        metrics.adClicks++;
        break;
      case UsageEventType.AD_VIDEO_START:
        metrics.videoStarts++;
        break;
      case UsageEventType.AD_VIDEO_COMPLETE:
        metrics.videoCompletes++;
        break;
      case UsageEventType.AD_REVENUE:
        if (event.revenueAmount != null) {
          metrics.totalRevenue += event.revenueAmount;
        }
        break;
      case UsageEventType.API_CALL:
        metrics.apiCalls++;
        break;
      case UsageEventType.CACHE_HIT:
        metrics.cacheHits++;
        break;
      case UsageEventType.CACHE_MISS:
        metrics.cacheMisses++;
        break;
      case UsageEventType.ERROR:
        metrics.errors++;
        break;
    }
  }

  // Convenience methods for recording events
  recordRequest(placementId: string, adapterId?: string, adFormat?: string): void {
    this.record({
      type: UsageEventType.AD_REQUEST,
      timestamp: Date.now(),
      placementId,
      adapterId,
      adFormat,
    });
  }

  recordImpression(placementId: string, adapterId: string, adFormat?: string): void {
    this.record({
      type: UsageEventType.AD_IMPRESSION,
      timestamp: Date.now(),
      placementId,
      adapterId,
      adFormat,
    });
  }

  recordClick(placementId: string, adapterId: string, adFormat?: string): void {
    this.record({
      type: UsageEventType.AD_CLICK,
      timestamp: Date.now(),
      placementId,
      adapterId,
      adFormat,
    });
  }

  recordRevenue(placementId: string, adapterId: string, amount: number): void {
    this.record({
      type: UsageEventType.AD_REVENUE,
      timestamp: Date.now(),
      placementId,
      adapterId,
      revenueAmount: amount,
    });
  }

  recordVideoStart(placementId: string, adapterId: string): void {
    this.record({
      type: UsageEventType.AD_VIDEO_START,
      timestamp: Date.now(),
      placementId,
      adapterId,
      adFormat: 'video',
    });
  }

  recordVideoComplete(placementId: string, adapterId: string): void {
    this.record({
      type: UsageEventType.AD_VIDEO_COMPLETE,
      timestamp: Date.now(),
      placementId,
      adapterId,
      adFormat: 'video',
    });
  }

  recordCacheHit(placementId?: string): void {
    this.record({
      type: UsageEventType.CACHE_HIT,
      timestamp: Date.now(),
      placementId,
    });
  }

  recordCacheMiss(placementId?: string): void {
    this.record({
      type: UsageEventType.CACHE_MISS,
      timestamp: Date.now(),
      placementId,
    });
  }

  recordError(placementId?: string, adapterId?: string): void {
    this.record({
      type: UsageEventType.ERROR,
      timestamp: Date.now(),
      placementId,
      adapterId,
    });
  }

  /**
   * Get current metrics snapshot
   */
  getMetrics(): UsageMetrics {
    return this.globalMetrics.toMetrics(this.periodStart, Date.now());
  }

  /**
   * Get breakdown by dimensions
   */
  getBreakdown(): UsageBreakdown {
    const now = Date.now();
    return {
      byPlacement: new Map(
        Array.from(this.placementMetrics.entries()).map(([k, v]) => [
          k,
          v.toMetrics(this.periodStart, now),
        ])
      ),
      byAdapter: new Map(
        Array.from(this.adapterMetrics.entries()).map(([k, v]) => [
          k,
          v.toMetrics(this.periodStart, now),
        ])
      ),
      byAdFormat: new Map(
        Array.from(this.formatMetrics.entries()).map(([k, v]) => [
          k,
          v.toMetrics(this.periodStart, now),
        ])
      ),
    };
  }

  /**
   * Get click-through rate
   */
  getCTR(): number {
    const metrics = this.getMetrics();
    if (metrics.adImpressions === 0) return 0;
    return metrics.adClicks / metrics.adImpressions;
  }

  /**
   * Get fill rate
   */
  getFillRate(): number {
    const metrics = this.getMetrics();
    if (metrics.adRequests === 0) return 0;
    return metrics.adImpressions / metrics.adRequests;
  }

  /**
   * Get video completion rate
   */
  getVideoCompletionRate(): number {
    const metrics = this.getMetrics();
    if (metrics.videoStarts === 0) return 0;
    return metrics.videoCompletes / metrics.videoStarts;
  }

  /**
   * Get cache hit rate
   */
  getCacheHitRate(): number {
    const metrics = this.getMetrics();
    const total = metrics.cacheHits + metrics.cacheMisses;
    if (total === 0) return 0;
    return metrics.cacheHits / total;
  }

  /**
   * Get effective CPM
   */
  getEffectiveCPM(): number {
    const metrics = this.getMetrics();
    if (metrics.adImpressions === 0) return 0;
    return (metrics.totalRevenue / metrics.adImpressions) * 1000;
  }

  /**
   * Flush metrics to reporter
   */
  async flush(): Promise<boolean> {
    if (!this.config.enableRemoteReporting) return true;

    const metrics = this.getMetrics();
    const breakdown = this.getBreakdown();

    try {
      if (this.reporter) {
        const success = await this.reporter.report(metrics, breakdown);
        if (success) {
          this.reset();
        }
        return success;
      }

      // Default fetch-based reporting
      const response = await fetch(this.config.reportEndpoint, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          metrics,
          breakdown: {
            byPlacement: Object.fromEntries(breakdown.byPlacement),
            byAdapter: Object.fromEntries(breakdown.byAdapter),
            byAdFormat: Object.fromEntries(breakdown.byAdFormat),
          },
        }),
      });

      if (response.ok) {
        this.reset();
        return true;
      }
      return false;
    } catch {
      return false;
    }
  }

  /**
   * Reset all counters
   */
  reset(): void {
    this.globalMetrics.reset();
    this.placementMetrics.forEach(m => m.reset());
    this.adapterMetrics.forEach(m => m.reset());
    this.formatMetrics.forEach(m => m.reset());
    this.pendingEvents = [];
    this.periodStart = Date.now();
  }

  /**
   * Export metrics as JSON
   */
  exportAsJSON(): string {
    const metrics = this.getMetrics();
    const breakdown = this.getBreakdown();

    return JSON.stringify(
      {
        metrics,
        computed: {
          ctr: this.getCTR(),
          fillRate: this.getFillRate(),
          videoCompletionRate: this.getVideoCompletionRate(),
          cacheHitRate: this.getCacheHitRate(),
          effectiveCPM: this.getEffectiveCPM(),
        },
        breakdown: {
          placementCount: breakdown.byPlacement.size,
          adapterCount: breakdown.byAdapter.size,
          formatCount: breakdown.byAdFormat.size,
        },
      },
      null,
      2
    );
  }

  /**
   * Get pending events count
   */
  getPendingEventsCount(): number {
    return this.pendingEvents.length;
  }
}

/**
 * Builder for UsageMeter
 */
export class UsageMeterBuilder {
  private config: MeteringConfig = {};
  private reporter?: MeteringReporter;

  withConfig(config: MeteringConfig): this {
    this.config = { ...this.config, ...config };
    return this;
  }

  withReporter(reporter: MeteringReporter): this {
    this.reporter = reporter;
    return this;
  }

  withFlushInterval(ms: number): this {
    this.config.flushIntervalMs = ms;
    return this;
  }

  withSamplingRate(rate: number): this {
    this.config.samplingRate = Math.max(0, Math.min(1, rate));
    return this;
  }

  withReportEndpoint(endpoint: string): this {
    this.config.reportEndpoint = endpoint;
    return this;
  }

  enableLocalStorage(enabled: boolean): this {
    this.config.enableLocalStorage = enabled;
    return this;
  }

  enableRemoteReporting(enabled: boolean): this {
    this.config.enableRemoteReporting = enabled;
    return this;
  }

  build(): UsageMeter {
    return new UsageMeter(this.config, this.reporter);
  }
}
